File: src/core-ts/rng.ts
/**
 * PCG32 — a small, fast, seedable PRNG (PCG-XSH-RR, 64-bit state / 32-bit output).
 *
 * This is the **shared reproducibility contract** for the whole project: the
 * WASM compute core (from M2) reimplements this exact algorithm so that, given
 * the same seed, the Rust core and this TS oracle produce byte-identical runs
 * (see PRD §9.2). MATLAB parity is only *statistical*, so we deliberately do NOT
 * reproduce MATLAB's Mersenne Twister — we define our own stream here and mirror
 * it in Rust.
 *
 * Reference: M.E. O'Neill, "PCG: A Family of Simple Fast Space-Efficient
 * Statistically Good Algorithms for Random Number Generation" (2014).
 *
 * TS uses BigInt for the 64-bit state (the oracle is not perf-critical). The
 * Rust port uses `u64`/`u32` wrapping arithmetic; every operation below maps
 * 1:1 onto wrapping ops, and the constants are identical.
 */

const MASK64 = (1n << 64n) - 1n;
const MASK32 = 0xffff_ffffn;
const MULTIPLIER = 6364136223846793005n;
const TWO_POW_32 = 1n << 32n;

/** Default stream selector (the canonical PCG default increment's `initseq`). */
const DEFAULT_STREAM = 0xda3e39cb94b95bdbn;

export class Pcg32 {
  private state = 0n;
  private readonly inc: bigint;

  /**
   * @param seed   initial state (`initstate`); any non-negative integer.
   * @param stream stream selector (`initseq`); different streams are distinct,
   *               non-overlapping sequences for the same seed.
   */
  constructor(seed: number | bigint, stream: number | bigint = DEFAULT_STREAM) {
    // pcg32_srandom_r: inc = (initseq << 1) | 1; state = 0; step; state += seed; step.
    this.inc = (((BigInt(stream) & MASK64) << 1n) | 1n) & MASK64;
    this.stepAndOutput();
    this.state = (this.state + (BigInt(seed) & MASK64)) & MASK64;
    this.stepAndOutput();
  }

  /** Advance the LCG state and emit one 32-bit output (PCG-XSH-RR). */
  private stepAndOutput(): number {
    const old = this.state;
    this.state = (old * MULTIPLIER + this.inc) & MASK64;
    const xorshifted = (((old >> 18n) ^ old) >> 27n) & MASK32;
    const rot = old >> 59n; // 0..31
    const result = ((xorshifted >> rot) | (xorshifted << (-rot & 31n))) & MASK32;
    return Number(result);
  }

  /** Next uniformly-distributed 32-bit unsigned integer, in `[0, 2^32)`. */
  nextU32(): number {
    return this.stepAndOutput();
  }

  /** Next float in `[0, 1)` with 32 bits of resolution. */
  nextFloat(): number {
    return this.nextU32() / 2 ** 32;
  }

  /**
   * Uniform integer in `[0, bound)`, unbiased via rejection (arc4random_uniform
   * style: reject the low `2^32 mod bound` outputs). Rust port computes the same
   * threshold with `(0u32.wrapping_sub(bound)) % bound`.
   */
  nextBounded(bound: number): number {
    if (bound <= 1) return 0;
    const b = BigInt(bound);
    const threshold = TWO_POW_32 % b;
    while (true) {
      const r = BigInt(this.nextU32());
      if (r >= threshold) return Number(r % b);
    }
  }

  /**
   * A random permutation of `[0, n)` via descending Fisher–Yates. This is our
   * deterministic replacement for MATLAB's `randperm`; the ordering differs from
   * MATLAB (statistical parity only) but is identical between TS and Rust.
   */
  permutation(n: number): Uint32Array {
    const out = new Uint32Array(n);
    for (let i = 0; i < n; i++) out[i] = i;
    for (let i = n - 1; i > 0; i--) {
      const j = this.nextBounded(i + 1);
      const tmp = out[i];
      out[i] = out[j];
      out[j] = tmp;
    }
    return out;
  }
}

File: core-wasm/src/rng.rs
//! PCG32 (PCG-XSH-RR, 64-bit state / 32-bit output).
//!
//! This is the Rust half of the shared reproducibility contract defined in
//! `src/core-ts/rng.ts`. Every operation mirrors the TS oracle exactly (the TS
//! side uses BigInt; here we use native `u64`/`u32` wrapping arithmetic), so a
//! given seed produces byte-identical runs across both. Validated against the
//! canonical PCG32 reference vector (seed 42, stream 54) in the unit tests.

const MULTIPLIER: u64 = 6364136223846793005;

/// Default stream selector — must equal `DEFAULT_STREAM` in `rng.ts`.
pub const DEFAULT_STREAM: u64 = 0xda3e39cb94b95bdb;

pub struct Pcg32 {
    state: u64,
    inc: u64,
}

impl Pcg32 {
    /// Seed with an explicit stream (`initstate` + `initseq`).
    pub fn new(seed: u64, stream: u64) -> Self {
        let mut rng = Pcg32 {
            state: 0,
            inc: (stream << 1) | 1,
        };
        rng.next_u32();
        rng.state = rng.state.wrapping_add(seed);
        rng.next_u32();
        rng
    }

    /// Seed using the default stream (matches `new Pcg32(seed)` in TS).
    pub fn with_seed(seed: u64) -> Self {
        Self::new(seed, DEFAULT_STREAM)
    }

    /// Next uniformly-distributed 32-bit unsigned integer.
    pub fn next_u32(&mut self) -> u32 {
        let old = self.state;
        self.state = old.wrapping_mul(MULTIPLIER).wrapping_add(self.inc);
        let xorshifted = (((old >> 18) ^ old) >> 27) as u32;
        let rot = (old >> 59) as u32;
        xorshifted.rotate_right(rot)
    }

    /// Next float in `[0, 1)` with 32 bits of resolution.
    pub fn next_f64(&mut self) -> f64 {
        f64::from(self.next_u32()) / 4294967296.0
    }

    /// Uniform integer in `[0, bound)`, unbiased via rejection.
    pub fn next_bounded(&mut self, bound: u32) -> u32 {
        if bound <= 1 {
            return 0;
        }
        let threshold = bound.wrapping_neg() % bound; // == 2^32 mod bound
        loop {
            let r = self.next_u32();
            if r >= threshold {
                return r % bound;
            }
        }
    }

    /// A random permutation of `[0, n)` via descending Fisher–Yates.
    pub fn permutation(&mut self, n: usize) -> Vec<u32> {
        let mut out: Vec<u32> = (0..n as u32).collect();
        for i in (1..n).rev() {
            let j = self.next_bounded(i as u32 + 1) as usize;
            out.swap(i, j);
        }
        out
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn matches_canonical_reference_vector() {
        // pcg32_srandom_r(&rng, 42, 54) — the published PCG32 demo output.
        let mut rng = Pcg32::new(42, 54);
        let expected: [u32; 6] = [
            0xa15c02b7, 0x7b47f409, 0xba1d3330, 0x83d2f293, 0xbfa4784b, 0xcbed606e,
        ];
        let got: Vec<u32> = (0..6).map(|_| rng.next_u32()).collect();
        assert_eq!(got, expected);
    }

    #[test]
    fn next_f64_in_unit_interval() {
        let mut rng = Pcg32::with_seed(7);
        for _ in 0..10_000 {
            let f = rng.next_f64();
            assert!((0.0..1.0).contains(&f));
        }
    }

    #[test]
    fn next_bounded_in_range_and_edge_cases() {
        let mut rng = Pcg32::with_seed(99);
        for _ in 0..10_000 {
            assert!(rng.next_bounded(7) < 7);
        }
        assert_eq!(rng.next_bounded(1), 0);
        assert_eq!(rng.next_bounded(0), 0);
    }

    #[test]
    fn permutation_is_valid() {
        let mut rng = Pcg32::with_seed(2024);
        let mut perm = rng.permutation(50);
        assert_eq!(perm.len(), 50);
        perm.sort_unstable();
        let expected: Vec<u32> = (0..50).collect();
        assert_eq!(perm, expected);
    }
}
